From 0f8f36d752e11e8bd5b1cb780efd68af12275202 Mon Sep 17 00:00:00 2001
From: Adrian DC <radian.dc@gmail.com>
Date: Mon, 14 Aug 2017 01:21:06 +0200
Subject: [PATCH 2/3] Adapt add_service uses for TARGET_HAS_LEGACY_CAMERA_HAL1

 * Avoid mutual neverallow between mediaserver and cameraserver
 * Add 'target_has_legacy_camera_hal1' variable to plat_policy.conf

Change-Id: I1f58f1144796601c85b743ce2cdcb97034b5301e
Signed-off-by: Adrian DC <radian.dc@gmail.com>
---
 Android.mk             | 1 +
 public/cameraserver.te | 8 +++++++-
 public/mediaserver.te  | 7 ++++++-
 3 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/Android.mk b/Android.mk
index bd714df..62a96c2 100644
--- a/Android.mk
+++ b/Android.mk
@@ -308,6 +308,7 @@ $(PLAT_PUBLIC_POLICY) $(PLAT_PRIVATE_POLICY))
 		-D target_arch=$(PRIVATE_TGT_ARCH) \
 		-D target_with_asan=$(PRIVATE_TGT_WITH_ASAN) \
 		-D target_full_treble=$(PRODUCT_FULL_TREBLE) \
+		-D target_has_legacy_camera_hal1=$(TARGET_HAS_LEGACY_CAMERA_HAL1) \
 		-D target_needs_platform_text_relocations=$(TARGET_NEEDS_PLATFORM_TEXT_RELOCATIONS) \
 		-D target_uses_legacy_adb_interface=$(TARGET_USES_LEGACY_ADB_INTERFACE) \
 		-s $^ > $@
diff --git a/public/cameraserver.te b/public/cameraserver.te
index 0dd4a80..8ededea 100644
--- a/public/cameraserver.te
+++ b/public/cameraserver.te
@@ -16,7 +16,13 @@ allow cameraserver ion_device:chr_file rw_file_perms;
 # Talk with graphics composer fences
 allow cameraserver hal_graphics_composer:fd use;
 
-add_service(cameraserver, cameraserver_service)
+ifelse(target_has_legacy_camera_hal1, `true',
+  allow cameraserver cameraserver_service:service_manager { add find };
+  neverallow { domain -cameraserver -mediaserver } cameraserver_service:service_manager add;
+,
+  add_service(cameraserver, cameraserver_service)
+)
+
 allow cameraserver appops_service:service_manager find;
 allow cameraserver audioserver_service:service_manager find;
 allow cameraserver batterystats_service:service_manager find;
diff --git a/public/mediaserver.te b/public/mediaserver.te
index cea3d86..074c8f6 100644
--- a/public/mediaserver.te
+++ b/public/mediaserver.te
@@ -75,7 +75,12 @@ unix_socket_connect(mediaserver, drmserver, drmserver)
 # but seems appropriate for all devices.
 unix_socket_connect(mediaserver, bluetooth, bluetooth)
 
-add_service(mediaserver, mediaserver_service)
+ifelse(target_has_legacy_camera_hal1, `true',
+  allow mediaserver mediaserver_service:service_manager { add find };
+  neverallow { domain -mediaserver -cameraserver } mediaserver_service:service_manager add;
+,
+  add_service(mediaserver, mediaserver_service)
+)
 allow mediaserver activity_service:service_manager find;
 allow mediaserver appops_service:service_manager find;
 allow mediaserver audioserver_service:service_manager find;
-- 
2.7.4

